control-plane/api-gateway/controllers/gateway_class_serializer.go

```diff
@@ -0,0 +1,91 @@
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/hashicorp/consul-k8s/control-plane/api/v1alpha1"
+)
+
+const (
+	Group               = "api-gateway.consul.hashicorp.com"
+	annotationConfigKey = "api-gateway.consul.hashicorp.com/config"
+)
+
+var ErrUnmarshallingGatewayClassConfig = errors.New("error unmarshalling GatewayClassConfig annotation, skipping")
+
+func SerializeGatewayClassConfig(ctx context.Context, client client.Client, gw *gwv1beta1.Gateway, gwc *gwv1beta1.GatewayClass) (bool, error) {
+	var (
+		config    v1alpha1.GatewayClassConfig
+		err       error
+		annotated bool
+		managed   bool
+	)
+
+	if gw.Annotations == nil {
+		gw.Annotations = make(map[string]string)
+	}
+
+	if annotatedConfig, ok := gw.Annotations[annotationConfigKey]; ok {
+		if err := json.Unmarshal([]byte(annotatedConfig), &config.Spec); err != nil {
+			return false, ErrUnmarshallingGatewayClassConfig
+		}
+		annotated = true
+		managed = true
+	}
+
+	if !managed {
+		// check if we own the gateway
+		config, managed, err = getConfigForGatewayClass(ctx, client, gwc)
+		if err != nil {
+			gw.Annotations[annotationConfigKey] = ""
+			if k8serrors.IsNotFound(err) {
+				// invalid config which means an invalid gatewayclass
+				// so pretend we don't exist
+				return false, nil
+			}
+			return false, err
+		}
+		if !managed {
+			gw.Annotations[annotationConfigKey] = ""
+			// we don't own this gateway so we pretend it doesn't exist
+			return false, nil
+		}
+	}
+
+	marshaled, err := json.Marshal(config.Spec)
+	if err != nil {
+		return false, err
+	}
+
+	gw.Annotations[annotationConfigKey] = string(marshaled)
+
+	if !annotated {
+		// we annotated for the first time
+		return true, client.Update(ctx, gw)
+	}
+
+	return false, nil
+}
```

Comment:
This is heavily adapted from https://github.com/hashicorp/consul-api-gateway/blob/main/internal/k8s/reconciler/manager.go#L127